pkg/deployment/ipam.go

```diff
@@ -0,0 +1,242 @@
+	err := reserveIPs(ctx, helper, instance, nodes, allIPSets)
+	if err != nil {
+		instance.Status.Conditions.MarkFalse(
+			dataplanev1beta1.RoleIPReservationReadyCondition,
+			condition.ErrorReason, condition.SeverityError,
+			dataplanev1beta1.RoleIPReservationReadyErrorMessage)
+		return false, err
+	}
+
+	if len(allIPSets) == 0 {
+		return true, nil
+	}
+
+	for _, s := range allIPSets {
+		if s.Status.Conditions.IsFalse(condition.ReadyCondition) {
+			return false, nil
+		}
+	}
+
+	instance.Status.Conditions.MarkTrue(
+		dataplanev1beta1.RoleIPReservationReadyCondition,
+		dataplanev1beta1.RoleIPReservationReadyMessage)
+	return true, nil
+
+}
+
+// createOrPatchDNSData builds the DNSData
+func createOrPatchDNSData(ctx context.Context, helper *helper.Helper,
+	instance *dataplanev1beta1.OpenStackDataPlaneRole,
+	nodes *dataplanev1beta1.OpenStackDataPlaneNodeList) error {
+
+	var allDNSRecords []networkv1beta1.DNSHost
+	// Build DNSData CR
+	for _, node := range nodes.Items {
+		nets := node.Spec.Node.Networks
+		if len(nets) == 0 {
+			nets = instance.Spec.NodeTemplate.Networks
+		}
+		if len(nets) > 0 {
+			// Get IPSet
+			ipSet := &networkv1beta1.IPSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      node.Name,
+					Namespace: instance.Namespace,
+				},
+			}
+			hostName := node.Spec.HostName
+			if hostName == "" {
+				hostName = node.Name
+
+			}
+			key := client.ObjectKeyFromObject(ipSet)
+			err := helper.GetClient().Get(ctx, key, ipSet)
+			if err == nil || !k8s_errors.IsNotFound(err) {
+				for _, res := range ipSet.Status.Reservation {
+					dnsRecord := networkv1beta1.DNSHost{}
+					dnsRecord.IP = res.Address
+					var fqdnNames []string
+					// Add domain once IPset status includes it
+					fqdnName := strings.Join([]string{hostName,
+						strings.ToLower(string(res.Network))}, ".")
+					fqdnNames = append(fqdnNames, fqdnName)
+					dnsRecord.Hostnames = fqdnNames
+					allDNSRecords = append(allDNSRecords, dnsRecord)
+				}
+			}
+		}
+	}
+	util.LogForObject(helper, "Reconciling DNSData", instance)
+	dnsData := &networkv1beta1.DNSData{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: instance.Namespace,
+			Name:      instance.Name,
+		},
+	}
+	_, err := controllerutil.CreateOrPatch(ctx, helper.GetClient(), dnsData, func() error {
+		dnsData.Spec.Hosts = allDNSRecords
+		// Set controller reference to the DataPlaneNode object
+		err := controllerutil.SetControllerReference(
+			helper.GetBeforeObject(), dnsData, helper.GetScheme())
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+
+}
+
+// EnsureDNSData Ensures DNSData is created
+func EnsureDNSData(ctx context.Context, helper *helper.Helper,
+	instance *dataplanev1beta1.OpenStackDataPlaneRole,
+	nodes *dataplanev1beta1.OpenStackDataPlaneNodeList) (bool, error) {
+	// Verify dnsmasq CR exists
+	dnsmasq := &networkv1beta1.DNSMasq{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "dnsmasq",
```

Comment:
We use that for all services right (I had earlier checked for the list but changed looking at other services) https://github.com/openstack-k8s-operators/openstack-operator/blob/main/pkg/openstack/cinder.go#L22

Reply:
for envs deployed with openstack-operator, we could set it to what we create it there, but if you deploy the operators individually, I think it would be better to do the list

Reply:
Do you expect individual operators deployed with dataplane?  Closing the discussion here I will revert to checking list as before.

Reply:
done!